### PR TITLE
Implementation of HD44780 LCD controller

### PR DESCRIPTION
**Pull request by k-brk**
Hi, I have been working lately on driver for HD44780 LCD controller and decided to share it. It supports LCDs that are up to 40 characters per line and up to 4 lines(except for 40x4 atm.), which covers most of HD44780 based displays. The driver is not finished yet, there is still some work to do, like some recovery action on too long active busy flag(In case display stucks on instruction somehow) or writing to DDRAM. Other than that implementation covers most of Hitachi hd44780 controller and compatible ones. Also I would like to provide support for I2C version(Its normal parallel 8bit + I/O i2c expander, usually in 20x4 LCDs, should't be much to do) as soon as it will be delivered.

Feedback appreciated.

### CODE REVIEW

**Comment by deadprogram** [comment]
Hi @k-brk this is very cool looking, thanks for working on it.

Do you think you could please add "smoke tests" like these for both of your examples? https://github.com/tinygo-org/drivers/blob/master/.circleci/config.yml#L16

It is far from perfect, but at least it ensures that all of the examples are buildable.

Perhaps @conejoninja might have some feedback, since he has done a bunch of display implementation?

File: hd44780/gpio.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+// SetCommandMode sets command/instruction mode
+func (g *GPIO) SetCommandMode(set bool) {
+	if set {
+		g.rs.Low()
+	} else {
+		g.rs.High()
+	}
+}
+
+func (g *GPIO) Write(data []byte) (n int, err error) {
```
Comment (conejoninja) [inline hd44780/gpio.go]: if it's exported, it should be docummented, Read function on this file too

File: hd44780/hd44780.go [hunk 2]
```diff
@@ -3,176 +3,45 @@ package hd44780
...
-	for i := uint8(0); i < uint8(len(g.dataPins)); i++ {
-		if g.dataPins[i].Get() {
-			bits |= (1 << i)
-		}
-	}
-	return bits
+	return newGPIO(dataPins, e, rs, rw, DATA_LENGTH_8BIT), nil
 }
 
 type Device struct {
```
Comment (conejoninja) [inline hd44780/hd44780.go]: I like to put all struct definitions on the beginning of the file, or at least before the functions that use it

File: hd44780/hd44780.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+
+// Busy returns true when hd447890 is busy
+func (d *Device) Busy() bool {
+	status := d.bus.Read()
+	return (status & BUSY) > 0
+}
+
+// SetPixel is not supported on devices which uses HD44780 driver
+func (d *Device) SetPixel(x, y int16, c color.RGBA) {
+	panic("HD44780 does not support setting individual pixels")
```
Comment (conejoninja) [inline hd44780/hd44780.go]: if you have an internal buffer, why is it not possible to set a single pixel?

File: examples/hd44780/customchar/main.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"machine"
+
+	"github.com/tinygo-org/drivers/hd44780"
+)
+
+func main() {
+
+	lcd, _ := hd44780.NewGPIO4Bit([]uint8{machine.P0, machine.P1, machine.P2, machine.P3}, machine.P4, machine.P5, machine.P6)
```
Comment (conejoninja) [inline examples/hd44780/customchar/main.go]: Maybe put each param in a new line to make it more clear to read, and also for the configuration below (same for the other example)

File: hd44780/gpio.go (hunk 1 above)
Comment (k-brk) [reply hd44780/gpio.go]: Done, although I think I should make GPIO unexported, it should not be used directly

**Comment by deadprogram** [comment]
Are there any other changes pending on this PR, or is it ready for merge?

**Comment by conejoninja** [comment]
I think everything is resolved and it's ready to be merged

**Comment by k-brk** [comment]
It's ready. There is also I2C version on market ( Normal parallel 8bit + pcf8574 IO expander soldered together), I will provide implementation probably within a month depending on aliexpress delivery and exams at university. I will create PR when I2C it's ready.

**Comment by deadprogram** [comment]
Thanks @k-brk for the contribution and @conejoninja for helping review. Now merging.

**Comment by prlarsen** [comment]
> It's ready. There is also I2C version on market ( Normal parallel 8bit + pcf8574 IO expander soldered together), I will provide implementation probably within a month depending on aliexpress delivery and exams at university. I will create PR when I2C it's ready.

Hi, is there a driver for PCF8574 yet? I had a couple of these 16x2 displays lying around and it would be cool to use them with TinyGo.

**Pull request merged**